### DISCUSSION TITLE
react-i18next integration examples

### DISCUSSION THREAD

**Discussion by crummy** [vikejs/vike#809] [Discussion] [answered]
Hi, I'm trying to integrate react-i18n into a VPS project, but I'm new to SSR and a little unsure of how to get it working. I've followed the existing i18n guide and got routing working fine, locale injected, but am struggling with the plugin integration.

My PageShell.tsx looks like this:

```
i18n.use(initReactI18next) // passes i18n down to react-i18next
  .use(LanguageDetector)
  .use(resourcesToBackend((language, ns) => import(`../../locales/${language}/${ns}.json`)))
  .init({
    fallbackLng: "en",
    lng: 'en',
    debug: true,

    interpolation: {
      escapeValue: false // react already safes from xss => https://www.i18next.com/translation-function/interpolation#unescape
    }
  });


export default function (props: Props & ServerContext) {
  return <BaseContext.Provider value={props}>
    <I18nextProvider i18n={i18n}>
      <Navbar url={props.urlOriginal}/>
      {props.children}
      <Footer/>
    </I18nextProvider>
  </BaseContext.Provider>
}
```

However, I get this error on load in my console:

```
i18next::translator: key "About Us" for languages "en" won't get resolved as namespace "translation" was not yet loaded This means something IS WRONG in your setup. You access the t function before i18next.init / i18next.loadNamespace / i18next.changeLanguage was done. Wait for the callback or Promise to resolve before accessing it!!! 
```

I suppose I need to `await` my i18n before I give to the provider, which I imagine I should be able to do server side on render rather than page load, but not sure about frontend... Is there an example somewhere of how this might work?

(If this question should be aimed on the react-i18next repo then let me know.)

**Comment by brillout** [comment, marked as answer]
Both `render()` hooks in `.page.client.js` and `.page.server.js` can be async. So, in principle, you should be able to await i18n init over there.

**Reply by brillout** [reply]
You can start the i18n init at server/browser start in the global scope of `renderPage()` and `.page.client.js`, then make sure to await the init before you render any page. AFAICT that'll give you optimal performance.

**Reply by crummy** [reply]
Thanks for the pointers. I found this helpful post as well: https://github.com/i18next/react-i18next/issues/819#issuecomment-1089898976

This allows me to skip the promise entirely.

**Reply by crummy** [reply]
(For anyone else working with VPS and react-i18next, I have an open question on SO here. https://stackoverflow.com/questions/76087180/adding-react-i18next-to-vite-plugin-ssr-project-i18next-was-not-initialized)

**Reply by brillout** [reply]
Thanks for circling back on this!
